### PR TITLE
[WIP] Make material changes interactive

### PR DESCRIPTION
**Pull request by usakhelo**
Hello,
This change adds the ability to change object's material interactively during Active Shade rendering mode. Please note changes to existing materials are reflected in active shade rendering. Applying new materials requires restarting of rendering.

Thanks for review!

### CODE REVIEW

**Comment by Mango-3** [comment]
A general comment:
Since Max 2019 update 3 Max has the possibility to interact/edit and render simultaneously in the same ActiveShade viewport. The SDK has for this a new `IActiveShadeFragmentManager` class: [http://help.autodesk.com/view/3DSMAX/2020/ENU/?guid=__developer_what_s_new_what_s_new_3ds_max_2020_sdk_html#rendering](url) &
[http://help.autodesk.com/view/3DSMAX/2020/ENU/?guid=__cpp_ref_class_max_s_d_k_1_1_i_active_shade_fragment_manager_html](url)

**Comment by usakhelo** [comment]
Thanks @Mango-3 , I didn't know about this! This is so cool!
Actually starting 2017 they changed interactive renderer so that some things like material changes handled by SDK. But we already have our own logic built so I thought rather than rewrite whole interactive renderer I'll just add material changes.

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 1]
```diff
@@ -220,8 +222,92 @@ namespace
...
+            }
+        }
+
+        void ControllerOtherEvent(NodeKeyTab & nodes) override 
+        {
+          for (int i = 0, e = nodes.Count(); i < e; ++i)
+          {
+            auto test = NodeEventNamespace::GetNodeByKey(nodes[i]);
+          }
+        };
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: Remove semicolon.

Insert blank line between methods.

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 2]
```diff
@@ -220,8 +222,92 @@ namespace
...
+                INode* node = NodeEventNamespace::GetNodeByKey(nodes[i]);
+                Mtl* mtl = node->GetMtl();
+                if (mtl == nullptr)
+                    continue;
+                m_renderer->update_material(mtl);
+                m_renderer->get_render_session()->reininitialize_render();
+            }
+        }
+
+        void ControllerOtherEvent(NodeKeyTab & nodes) override
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: Remove space before `&` (fix everywhere).

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 3]
```diff
@@ -220,8 +222,92 @@ namespace
                     break;
                 }
             }
+
+            //TODO: update lights here
+        }
+        
+        void MaterialStructured(NodeKeyTab& nodes)
+        {
+            if (m_renderer == nullptr)
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: Is `m_renderer == nullptr` something that can normally happen, or would it be a programmer error? If the former, keep the code as is. If the latter, replace by an assertion.

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 4]
```diff
@@ -220,8 +222,92 @@ namespace
...
+                return;
+
+            for (int i = 0, e = nodes.Count(); i < e; ++i)
+            {
+                INode* node = NodeEventNamespace::GetNodeByKey(nodes[i]);
+                Mtl* mtl = node->GetMtl();
+                if (mtl == nullptr)
+                    continue;
+                m_renderer->assign_material(mtl, node);
+                m_renderer->get_render_session()->reininitialize_render();
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: Is it OK to reinitialize the render at every iteration of the loop, or should it be done once outside the loop? Just asking.

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 5]
```diff
@@ -220,8 +222,92 @@ namespace
...
+                m_renderer->update_material(mtl);
+                m_renderer->get_render_session()->reininitialize_render();
+            }
+        }
+
+        void ControllerOtherEvent(NodeKeyTab & nodes) override 
+        {
+          for (int i = 0, e = nodes.Count(); i < e; ++i)
+          {
+            auto test = NodeEventNamespace::GetNodeByKey(nodes[i]);
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: This code doesn't do anything (same for all other methods). Remove?

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 6]
```diff
@@ -371,13 +458,89 @@ asf::auto_release_ptr<asr::Project> AppleseedInteractiveRender::prepare_project(
...
+            materials.push_back(submtl);
+        }
+    }
+    else
+    {
+        materials.push_back(mtl);
+    }
+
+    MaterialMap assigned_materials;
+    for (const auto& mtl : materials)
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: Specify the type instead of using `auto`, it makes the code easier to follow.

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 7]
```diff
@@ -371,13 +458,89 @@ asf::auto_release_ptr<asr::Project> AppleseedInteractiveRender::prepare_project(
...
+    {
+        IAppleseedMtl* appleseed_mtl =
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: It looks like you don't need that variable, and you don't need the cast either. Maybe this would be sufficient:

    if (mtl->GetInterface(IAppleseedMtl::interface_id()) == nullptr)
        continue;

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 8]
```diff
@@ -371,13 +458,89 @@ asf::auto_release_ptr<asr::Project> AppleseedInteractiveRender::prepare_project(
...
+    }
+    
+    auto instance_map = m_object_instance_map.find(node->GetObjectRef());
+    if (instance_map != m_object_instance_map.end())
+        get_render_session()->schedule_assign_material(assigned_materials, instance_map->second);
+}
+
+void AppleseedInteractiveRender::update_material(Mtl* mtl)
+{
+    if (mtl == nullptr)
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: Is it legal to call `update_material()` with a null argument? If not then replace this by an assertion.

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp [hunk 9]
```diff
@@ -371,13 +458,89 @@ asf::auto_release_ptr<asr::Project> AppleseedInteractiveRender::prepare_project(
...
+        IAppleseedMtl* appleseed_mtl =
+            static_cast<IAppleseedMtl*>(mtl->GetInterface(IAppleseedMtl::interface_id()));
+        if (appleseed_mtl == nullptr)
+            continue;
+
+        auto it = m_material_map.find(mtl);
+        if (it == m_material_map.end())
+            continue;
+
+        updated_materials[appleseed_mtl] = it->second;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.cpp]: You're using a mixture of `mtl` (in the `find()` call) and `appleseed_mtl`. Is that on purpose? Can this code be refactored to always use just one?

File: src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.h [hunk 10]
```diff
@@ -95,6 +96,8 @@ class AppleseedInteractiveRender
     void AbortRender() override;
 
     void update_camera_object(INode* camera);
+    void assign_material(Mtl* mtl, INode* node);
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/appleseedinteractive.h]: If `node` is the "receiver" of the material, I would suggest to put it in first position.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h [hunk 11]
```diff
@@ -80,6 +85,106 @@ class CameraObjectUpdateAction
...
+        for (const auto& mtl : m_material_map)
+        {
+            renderer::Material* material = assembly->materials().get_by_name(mtl.second.c_str());
+
+            if (material)
+                assembly->materials().remove(material);
+
+            assembly->materials().insert(
+                mtl.first->create_material(
+                *assembly,
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h]: Indent.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h [hunk 12]
```diff
@@ -80,6 +85,106 @@ class CameraObjectUpdateAction
...
+                    make_unique_name(assembly->materials(), wide_to_utf8(mtl.first->GetName()) + "_mat");
+
+                IAppleseedMtl* appleseed_mtl =
+                    static_cast<IAppleseedMtl*>(mtl.first->GetInterface(IAppleseedMtl::interface_id()));
+                if (appleseed_mtl == nullptr)
+                    continue;
+
+                const size_t index = assembly->materials().insert(
+                    appleseed_mtl->create_material(
+                    *assembly,
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h]: Indent.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h [hunk 13]
```diff
@@ -80,6 +85,106 @@ class CameraObjectUpdateAction
...
+        renderer::Project&      project,
+        const IAppleseedMtlMap& material_map)
+      : m_material_map(material_map)
+      , m_project(project)
+    {
+    }
+
+    void update() override
+    {
+        renderer::Assembly* assembly = m_project.get_scene()->assemblies().get_by_name("assembly");
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h]: Assert that `assembly` isn't `nullptr`.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h [hunk 14]
```diff
@@ -80,6 +85,106 @@ class CameraObjectUpdateAction
...
+                if (appleseed_mtl == nullptr)
+                    continue;
+
+                const size_t index = assembly->materials().insert(
+                    appleseed_mtl->create_material(
+                    *assembly,
+                    mtl_name.c_str(),
+                    false,
+                    GetCOREInterface()->GetTime()));
+                material = assembly->materials().get_by_index(index);
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h]: You can simply get the pointer before inserting the material.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h [hunk 15]
```diff
@@ -80,6 +85,106 @@ class CameraObjectUpdateAction
...
+            }
+
+            for (const auto& obj_instance : m_instances)
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h]: Replace `auto` by real type.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h [hunk 16]
```diff
@@ -80,6 +85,106 @@ class CameraObjectUpdateAction
...
+            {
+                auto* instance = obj_instance.first;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h]: Replace `auto` by real type.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h [hunk 17]
```diff
@@ -80,6 +85,106 @@ class CameraObjectUpdateAction
...
+                auto& obj_info = obj_instance.second;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h]: Replace `auto` by real type.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h [hunk 18]
```diff
@@ -80,6 +85,106 @@ class CameraObjectUpdateAction
...
+
+                auto front_mapping = instance->get_front_material_mappings();
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h]: Replace `auto` by real type.

Keep a const ref instead of making a copy (if possible; I don't know).

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h [hunk 19]
```diff
@@ -80,6 +85,106 @@ class CameraObjectUpdateAction
...
+                for (const auto& mat_mapping : front_mapping)
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h]: Replace `auto` by real type.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h [hunk 20]
```diff
@@ -80,6 +85,106 @@ class CameraObjectUpdateAction
...
+                {
+                    // TODO: recognize multi materials
+                    auto test1 = mat_mapping.key();
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h]: Remove.

File: src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h [hunk 21]
```diff
@@ -80,6 +85,106 @@ class CameraObjectUpdateAction
...
+                    auto test2 = mat_mapping.value();
+                    instance->assign_material(mat_mapping.key(), renderer::ObjectInstance::FrontSide, material->get_name());
+                    //front_material_mappings.insert(entry.second, material_info.m_name);
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h]: Remove.

File: src/appleseed-max-impl/appleseedinteractive/interactivesession.cpp [hunk 22]
```diff
@@ -105,3 +108,16 @@ void InteractiveSession::schedule_camera_update(
...
 }
+
+void InteractiveSession::schedule_material_update(const IAppleseedMtlMap& material_map)
+{
+    m_renderer_controller->schedule_update(
+        std::unique_ptr<ScheduledAction>(new MaterialUpdateAction(*m_project, material_map)));
+}
+
+void InteractiveSession::schedule_assign_material(
+    const MaterialMap& material_map, const InstanceMap& instances)
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactivesession.cpp]: One arg per line.

File: src/appleseed-max-impl/appleseedinteractive/interactivesession.h [hunk 23]
```diff
@@ -66,6 +68,10 @@ class InteractiveSession
 
     void schedule_camera_update(
         foundation::auto_release_ptr<renderer::Camera>  camera);
+    void schedule_material_update(
+        const IAppleseedMtlMap& material_map);
+    void schedule_assign_material(
+        const MaterialMap& material_map, const InstanceMap& instances);
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedinteractive/interactivesession.h]: One arg per line.

File: src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp [hunk 24]
```diff
@@ -141,11 +140,11 @@ namespace
     }
 
     std::string insert_default_material(
-        asr::Assembly&          assembly,
+        asr::Assembly*          assembly,
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp]: Why this change? `assembly` should never be `nullptr` so a ref is better. Revert.

File: src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp [hunk 25]
```diff
@@ -834,19 +845,22 @@ namespace
...
                 instance_name.c_str(),
                 params,
                 object_info.m_name.c_str(),
                 transform,
                 front_material_mappings,
                 back_material_mappings));
+
+        auto* object_instance = assembly.object_instances().get_by_index(instance_index);
+        instance_map.insert(std::make_pair(
+            assembly.object_instances().get_by_index(instance_index),
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp]: That's `object_instance`.

File: src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp [hunk 26]
```diff
@@ -856,12 +870,15 @@ namespace
...
         const TimeValue         time,
         ObjectMap&              object_map,
+        ObjectInstanceMap&      obj_instance_map,
         MaterialMap&            material_map,
         AssemblyMap&            assembly_map)
     {
         // Retrieve the geometrical object referenced by this node.
         Object* object = node->GetObjectRef();
 
+        std::map<asr::ObjectInstance*, ObjectInfo> instance_map;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp]: What's the difference between `obj_instance_map` (passed in argument) and `instance_map`?

File: src/appleseed-max-impl/appleseedrenderer/projectbuilder.h [hunk 27]
```diff
@@ -42,18 +42,34 @@
 
 // Standard headers.
 #include <vector>
+#include <map>
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/projectbuilder.h]: Order alphabetically.

File: src/appleseed-max-impl/appleseedrenderer/projectbuilder.h [hunk 28]
```diff
@@ -42,18 +42,34 @@
...
 
 // Forward declarations.
 namespace renderer { class Camera; }
+namespace renderer { class ObjectInstance; }
 namespace renderer { class ParamArray; }
 namespace renderer { class Project; }
+class IAppleseedMtl;
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/projectbuilder.h]: Order alphabetically.

**Comment by usakhelo** [comment]
Hi @dictoon,
Sorry, I was working on this change for sometime without updating the PR. Some of the changes here are irrelevant by now, but I'll fix your comments and then push couple of commits or probably do another PR.
